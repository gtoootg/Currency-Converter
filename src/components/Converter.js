import axios from 'axios'
import { useEffect, useState } from 'react';
import Header from './Header';

export default function Converter() {
  const [currencyList, setCurrencyList] = useState({})
  const [exchangeRates,setExchangeRates] = useState ({})
  const [upperCurrency, setUpperCurrency] = useState('EUR')
  const [lowerCurrency, setLowerCurrency] = useState('USD')
  const [upperCurrencyValue, setUpperCurrencyValue] = useState(1)
  const [lowerCurrencyValue, setLowerCurrencyValue] = useState()

  const [RegexErrorUpper,setRegexErrorUpper] = useState(false)
  const [RegexErrorLower,setRegexErrorLower] = useState(false)
 
  function sortObject(object){
    let array = Object.entries(object)
    array.sort(function(p1, p2){
      var p1Key = p1[0], p2Key = p2[0];
      if(p1Key < p2Key){ return -1; }
      if(p1Key > p2Key){ return 1; }
      return 0;
    })
    const sortedObject = Object.fromEntries(array);
    return sortedObject
  }


  function fetchCurrencyList(){
    axios.get('https://openexchangerates.org/api/currencies.json')
    .then((res)=>{
      setCurrencyList(sortObject(res.data))
    })

    axios.get('http://api.exchangeratesapi.io/v1/latest?access_key=3298054fd6cea8d6cbc4f26298ad9f2d')
    .then((res)=>{
      setExchangeRates(res.data.rates)
    })
  }

  function DisplayCurrencyList({initialCurrency,currentCurrency,setCurrency}){
    return(
      <select 
        onChange={(e)=>{
          e.preventDefault()
          setCurrency(e.target.value)}}
        value={currentCurrency}
      >
        <option value={initialCurrency} selected>{currencyList[initialCurrency]}</option>
        {Object.keys(currencyList).map((list,i)=>
            <option 
              key={i}
              value={list}
            >
              {currencyList[list]}
            </option>
          )}
      </select>
    )
  }

  // function checkNumber(val){
  //   const reg = new RegExp(/^[0-9,]*$/)
  //   return reg.test(val)
  // }

  async function convertCurrency(from, valueFrom, to, setValueTo){
    // await setRegexErrorUpper(!checkNumber(upperCurrencyValue))
    // await setRegexErrorLower(!checkNumber(lowerCurrencyValue))
    // const formatNumber= (number)=>{
    //   return Intl.NumberFormat('de-DE',{style:'decimal'}).format(number)
    // }
 
    const rateFrom = exchangeRates[from]
    const rateTo = exchangeRates[to]

    setValueTo(valueFrom * rateFrom / rateTo)
    // setValueTo(formatNumber(valueFrom)*formatNumber(rateFrom)/formatNumber(rateTo))
  }

  useEffect(()=>{
    convertCurrency( upperCurrency, upperCurrencyValue, lowerCurrency, setLowerCurrencyValue)
  },[upperCurrencyValue,upperCurrency])

  useEffect(()=>{
    convertCurrency( lowerCurrency, lowerCurrencyValue, upperCurrency, setUpperCurrencyValue)
  },[lowerCurrencyValue,lowerCurrency])

  useEffect(()=>{
    fetchCurrencyList()

  },[])


  function InputField(){
    return(
      <>
        <div>
          <input
            value={Intl.NumberFormat('de-DE',{style:'decimal'}).format(upperCurrencyValue)}
            onChange={async (e)=>{
              setUpperCurrencyValue(e.target.value)
            }}    
          />
          <DisplayCurrencyList
            initialCurrency='EUR'
            currentCurrency={upperCurrency}
            setCurrency={setUpperCurrency}
          />
          {RegexErrorUpper? <span>Please Input in a correct format</span>:null}
        </div>
        <div>
          <span>{lowerCurrency} : {lowerCurrencyValue}</span>
          <br/>
          <input
            value={Intl.NumberFormat('de-DE',{style:'decimal'}).format(lowerCurrencyValue)}
            onChange={(e)=>{
              setLowerCurrencyValue(e.target.value)
            }}    
          />
          <DisplayCurrencyList
            initialCurrency='USD'
            currentCurrency={lowerCurrency}
            setCurrency={setLowerCurrency}
          />
          {RegexErrorLower? <span>Please Input in a correct format</span>:null}
        </div>
      </>
    )
  }

  return (
    <>
      <Header
        upperCurrency={upperCurrency}
        upperCurrencyValue={upperCurrencyValue}
        lowerCurrency={lowerCurrency}
        lowerCurrencyValue={lowerCurrencyValue}
      />
      <InputField/>
    </>
  );
}
