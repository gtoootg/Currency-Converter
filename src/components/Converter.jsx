import axios from 'axios'
import React, { useEffect, useState } from 'react'


function InputField({currencyValue,setCurrencyValue}){
  return(
    <input
      data-testid="input"
      value={currencyValue? Intl.NumberFormat('de-DE',{style:'decimal'}).format(currencyValue):null}
      // value={currencyValue? currencyValue : null}
      onChange={async (e)=>{
        e.preventDefault()
        setCurrencyValue(e.target.value)
      }}    
  />
  )
}

export default function Converter(prop) {

  const {
    currencyList, 
    setCurrencyList,
    exchangeRates,
    setExchangeRates,
    upperCurrency, 
    setUpperCurrency,
    lowerCurrency, 
    setLowerCurrency,
    upperCurrencyValue, 
    setUpperCurrencyValue,
    lowerCurrencyValue, 
    setLowerCurrencyValue
  } = prop
  const [regexError, setRegexError] = useState(false)
 
  function sortObject(object){
    let array = Object.entries(object)
    array.sort(function(p1, p2){
      var p1Key = p1[0], p2Key = p2[0];
      if(p1Key < p2Key){ return -1; }
      if(p1Key > p2Key){ return 1; }
      return 0;
    })
    const sortedObject = Object.fromEntries(array);
    return sortedObject
  }


  function fetchCurrencyList(){
    axios.get('https://openexchangerates.org/api/currencies.json')
    .then((res)=>{
      setCurrencyList(sortObject(res.data))
    })

    axios.get('http://api.exchangeratesapi.io/v1/latest?access_key=3298054fd6cea8d6cbc4f26298ad9f2d')
    .then((res)=>{
      setExchangeRates(res.data.rates)
    })
  }

  function DisplayCurrencyList({initialCurrency,currentCurrency,setCurrency}){
    return(
      <select 
        onChange={(e)=>{
          e.preventDefault()
          setCurrency(e.target.value)}}
        value={currentCurrency}
        data-testid="select"
      >
        <option value={initialCurrency} selected>{currencyList[initialCurrency]}</option>
        {Object.keys(currencyList).map((list,i)=>
            <option 
              key={i}
              value={list}
              data-testid={list}
            >
              {currencyList[list]}
            </option>
          )}
      </select>
    )
  }

  function checkNumber(val){
    const reg = new RegExp(/^[0-9,.]*$/)
    if(!val) {return false}
    return reg.test(val)
  }

  async function convertCurrency(from, valueFrom, to, setValueTo){
    // await setRegexError(!checkNumber(valueFrom))
    // await setRegexErrorLower(!checkNumber(lowerCurrencyValue))
    // const formatNumber= (number)=>{
    //   return Intl.NumberFormat('de-DE',{style:'decimal'}).format(number)
    // }
 
    const rateFrom = exchangeRates[from]
    const rateTo = exchangeRates[to]

    setValueTo(valueFrom * rateFrom / rateTo)
    // setValueTo(formatNumber(valueFrom)*formatNumber(rateFrom)/formatNumber(rateTo))
  }

  useEffect(()=>{
    convertCurrency( upperCurrency, upperCurrencyValue, lowerCurrency, setLowerCurrencyValue)
  },[upperCurrencyValue,upperCurrency])

  useEffect(()=>{
    convertCurrency( lowerCurrency, lowerCurrencyValue, upperCurrency, setUpperCurrencyValue)
  },[lowerCurrencyValue,lowerCurrency])

  useEffect(()=>{
    fetchCurrencyList()
  },[])

  return(
    <>
      <div>
        <InputField
          currencyValue={upperCurrencyValue}
          setCurrencyValue={setUpperCurrencyValue}
        />
        <DisplayCurrencyList
          initialCurrency='EUR'
          currentCurrency={upperCurrency}
          setCurrency={setUpperCurrency}
        />
      </div>
      <br/>
      <div>
        <InputField
          currencyValue={lowerCurrencyValue}
          setCurrencyValue={setLowerCurrencyValue}
        />
        <DisplayCurrencyList
          initialCurrency='USD'
          currentCurrency={lowerCurrency}
          setCurrency={setLowerCurrency}
        />
        {regexError? <span>Please Input in a correct format</span>:null}
      </div>
    </>
  )
  

}
