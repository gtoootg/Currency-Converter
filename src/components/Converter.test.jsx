/**
 * @jest-environment jsdom
 */
import axios from "axios";
import { render } from "@testing-library/react";
import Converter from "./Converter";


describe('Converter',()=>{

  const currencyMock={
    "EUR": "Euro",
    "USD": "United States Dollar",
  }

  const exchangeRatesMock = {
    "success":true,
    "timestamp":1654356843,
    "base":"EUR",
    "date":"2022-06-04",
    "rates":{
      "EUR":1,
      "USD":1.071983
    }
  }

  jest.mock("axios");
  axios.get
    .mockResolvedValueOnce(currencyMock)
    .mockResolvedValueOnce(exchangeRatesMock)

  it('should render correctly', ()=>{
    const {container} = render(<Converter/>)

    expect(container.childElementCount).toBe(2)
  })
})