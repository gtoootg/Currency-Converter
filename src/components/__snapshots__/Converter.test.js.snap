// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Converter should render correctly 1`] = `
<div>
  <div
    data-testid="Converter"
  >
    <div>
      <h1>
        EUR
         : 
      </h1>
      <h1
        style="color: blue;"
      >
        USD
         : 
      </h1>
    </div>
    <div>
      <input
        value="NaN"
      />
      <select>
        <option
          value="EUR"
        />
      </select>
    </div>
    <div>
      <span>
        USD
         : 
        NaN
      </span>
      <br />
      <input
        value="NaN"
      />
      <select>
        <option
          value="USD"
        />
      </select>
    </div>
  </div>
</div>
`;
